JP-3956: Fix error propagation for inverse steps (photom, flatfield)

pllim opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3956](https://jira.stsci.edu/browse/JP-3956)

<!-- If this PR closes a GitHub issue that is not attached to a Jira ticket, uncomment the line below, and reference it here by its number -->
<!-- Closes # -->

<!-- describe the changes comprising this PR here -->
This PR addresses:

* BUG: Fix err and var for inverse steps.
* STY: Minor docstring and style fixes.

Also see:

* #9330
* https://github.com/spacetelescope/jwst/pull/9085
* https://github.com/spacetelescope/jwst/pull/9254
* https://github.com/spacetelescope/jwst/pull/9320

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [ ] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md)) https://github.com/spacetelescope/RegressionTests/actions/runs/14477166563 and https://github.com/spacetelescope/RegressionTests/actions/runs/14479420182
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

pllim commented:
Sorry about all the unrelated fixes. I told myself to be good and only do the actual fix but then I was just too triggered.

pllim commented:
I also made changes to flatfield. There are actual differences seen in RT but I cannot tell if they are expected or not: https://github.com/spacetelescope/RegressionTests/actions/runs/14252934458/job/39949867173

SCI changed too but not by much and not a lot of pixels. I am guessing there are floating point differences between `data /= factor` and `data *= precomputed_inverse_factor`. 🤷‍♀️

melanieclarke commented:
> I also made changes to flatfield. There are actual differences seen in RT but I cannot tell if they are expected or not: https://github.com/spacetelescope/RegressionTests/actions/runs/14252934458/job/39949867173
> 
> SCI changed too but not by much and not a lot of pixels. I am guessing there are floating point differences between `data /= factor` and `data *= precomputed_inverse_factor`. 🤷‍♀️

That is more changes than I would have expected.  I agree they're probably not scientifically significant, but we'll need to look through them.  Can you try running a couple cases locally and see if `data /= factor` restores the truth values, for comparison?

This one definitely looked problematic though: `test_nirspec_mos_mbkg[masterbg2d]: jwst.regtest.test_nirspec_masterbackground`
There are some new `inf` values in the VAR_FLAT and ERR extensions.  This is the one place where the inverted error propagations will show up in the products, so probably we need to check for some edge cases in the new variance handling.  NIRSpec flat variance is prone to floating point overflow.

pllim commented:
> check for some edge cases in the new variance handling

Is there a standard way for such checks in this package I can refer to as example implementation? For instance, what do I do when a factor is zero and inversing it becomes INF? Do I also have to set DQ flag?

melanieclarke commented:
> Is there a standard way for such checks in this package I can refer to as example implementation? For instance, what do I do when a factor is zero and inversing it becomes INF? Do I also have to set DQ flag?

It depends on the operation.  If we can avoid overflow with a better order of operations, we should.  We also try to avoid inf if possible, by handling divide by zero cases separately -- NaN is often a more acceptable value than inf.

Here's an example from the flat_field step for handling both:
https://github.com/spacetelescope/jwst/blob/21bd3589d60dee1cc084d19ba60d4fb6a060fc16/jwst/flatfield/flat_field.py#L672
Since that's the module you're working in, maybe start with that as a model.

For DQ handling, generally we match NaN values across SCI, ERR, and VAR arrays and also set DQ = DO_NOT_USE.  This is handled by a util function:
https://github.com/spacetelescope/jwst/blob/21bd3589d60dee1cc084d19ba60d4fb6a060fc16/jwst/lib/pipe_utils.py#L89
It ought to be called at the end of any step that creates new NaNs, so you shouldn't need to explicitly worry about it in the processing, but it's a good idea to check.  For flat_field, there is a unit test to verify in the forward processing:
https://github.com/spacetelescope/jwst/blob/21bd3589d60dee1cc084d19ba60d4fb6a060fc16/jwst/flatfield/tests/test_flatfield.py#L67-L73
But we should probably add one for the inverse processing as well for verification. 

There are also a couple regtests that check this, e.g.:
https://github.com/spacetelescope/jwst/blob/21bd3589d60dee1cc084d19ba60d4fb6a060fc16/jwst/regtest/test_nirspec_steps_spec2.py#L61
If they doesn't cover the code you're modifying, we should make sure to write one that does.

pllim commented:
Thanks for the input! I might not get back to this until next week.

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,25 @@ def apply_flat_field(science, flat, inverse=False):
+    flat_data_squared = flat_data ** 2
     if not inverse:
         science.data /= flat_data
+        science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
```
I am trying to grok the math. In the original code, the math was

```python
science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
```

that also happens after

```python
science.data /= flat_data
```

Why divide by `flat_data_squared` again in `var_flat` calculation when `science.data` itself was already divided by `flat_data` in-place?

melanieclarke replied:
This is an error propagation thing, for the uncertainty on a quantity A/B when both A and B have associated errors. See for example: https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,25 @@ def apply_flat_field(science, flat, inverse=False):
     if not isinstance(science, datamodels.GuiderCalModel):
-        flat_data_squared = flat_data ** 2
-        science.var_poisson /= flat_data_squared
-        science.var_rnoise /= flat_data_squared
```
While I understand that Poisson noise follows data value, I also don't understand why `flat_data_squared` should be applied to read noise. Isn't read noise hardware related and is not affected by flatfield?

drlaw1558 replied:
The read noise itself is hardware related and doesn't 'see' the flatfield, but you can think of it as a unit conversion.  If the SNR of a pixel is 10 prior to flatfielding, you want to ensure that it's still 10 after application of the flatfield too, which requires applying the conversion factor to both numerator and denominator.  (Ignoring here the noise introduced by the flatfield step itself, which is a separate term).

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -419,17 +427,20 @@ def nirspec_fs_msa(output_model, f_flat_model, s_flat_model, d_flat_model, dispa
+            slit.var_poisson *= flat_data_squared
+            slit.var_rnoise *= flat_data_squared
+            slit.var_flat = (slit.data * slit_flat.data * slit_flat.err) ** 2
+            np.place(slit.var_flat, np.isinf(slit.var_flat), np.nan)
```
I think this `slit.var_flat` calculation caused INF in some of the regtest. I can turn it into NaN like this but I am not sure if it is correct way to handle. As Melanie pointed out, the NaN propagates and also then affects subsequent processing.

melanieclarke replied:
i.e.: `slit.var_flat = (slit.data * slit_flat.data / slit_flat.err) ** 2`

melanieclarke replied:
On inverse, we're trying to recover the original data and error values, as if the flat processing never happened.  For the error and variance, it's not a propagation process, it's just an undo.

There will be some pixels that are not invertible since the flat process introduces NaNs, and that's okay - we should just return NaNs there.  But for any values with sensible numbers we should aim for data == flat.inverse(flat(data)), error == flat.inverse(flat(error)), variance == flat.inverse(flat(variance)).  If that makes sense?

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,25 @@ def apply_flat_field(science, flat, inverse=False):
+        else:
+            science.var_poisson *= flat_data_squared
+            science.var_rnoise *= flat_data_squared
         science.err = np.sqrt(science.var_poisson + science.var_rnoise + science.var_flat)
```
When the ERR math is generic like this, I don't have to inverse anything, right?

melanieclarke replied:
Yes, that should be fine forwards or backwards.

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -419,17 +427,20 @@ def nirspec_fs_msa(output_model, f_flat_model, s_flat_model, d_flat_model, dispa
+            slit.var_poisson *= flat_data_squared
+            slit.var_rnoise *= flat_data_squared
+            slit.var_flat = (slit.data * slit_flat.data / slit_flat.err) ** 2
+            #np.place(slit.var_flat, np.isinf(slit.var_flat), np.nan)  # REMOVE?
```
Note to self: See if np.place is still needed after another RT. If not, remove from code.

pllim commented:
I think the INF is gone. Should I wait for some "okify" to go through first to make sure any RT failures are actually from this PR?

melanieclarke commented:
> I think the INF is gone. Should I wait for some "okify" to go through first to make sure any RT failures are actually from this PR?

I think you can go ahead and run them. I don't know of any outstanding failures in RT right now, except the 1 SOSS failure in devdeps.

pllim commented:
RT has 5 failures; same 5 for stable and devdeps. Who should review the diff? If these are okay, I will remove the commented out `np.place` hack and take this out of draft. Thanks for your patience!

![Screenshot 2025-04-09 183202](https://github.com/user-attachments/assets/2f9b7a06-6c1f-44e7-9f0e-bc69da2dbf81)

melanieclarke commented:
> RT has 5 failures; same 5 for stable and devdeps. Who should review the diff? If these are okay, I will remove the commented out `np.place` hack and take this out of draft. Thanks for your patience!

The changes to the error and variance in the masterbg2d product for MOS are the only changes I would expect to see.  The SOSS product changes are confusing -- I don't think they should be there.  When we are able to run RT on main again, we should check if there are diffs on main for those files.

pllim commented:
I was hoping I could hook into existing flatfield tests but turns out all the test data have unity flat, which would defeat the purpose of inverse. I will have to ponder writing a completely new test for this.

pllim commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -265,13 +265,23 @@ def test_nirspec_fs_flat(srctype):
-    # null flat, so data is the same, other than nan edges
     nn = ~np.isnan(result.slits[0].data)
-    assert np.allclose(result.slits[0].data[nn], data.slits[0].data[nn])
+    assert_allclose(result.slits[0].data[nn], 1.371742, rtol=1e-6)  # FIXME: should be orig_data.slits[0].data[nn] / 0.9 ?
```
I am stuck designing a test for the inverse roundtrip. Would need a flatfield step expert to help here.

pllim replied:
https://innerspace.stsci.edu/display/SCSB/JWST+calibration+pipeline+maintainers says @jemorrison @stscirij @melanieclarke ? 🙏

pllim replied:
Got it. Thanks!

pllim commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -252,9 +251,9 @@ def test_nirspec_fs_flat(srctype):
-    data.slits[0].err = np.full(shape, 0.0)
-    data.slits[0].var_poisson = np.full(shape, 0.0)
-    data.slits[0].var_rnoise = np.full(shape, 0.0)
+    data.slits[0].err = np.full(shape, 0.1)
+    data.slits[0].var_poisson = np.full(shape, 0.01)
+    data.slits[0].var_rnoise = np.full(shape, 0.01)
```
I don't remember why I changed these. But then again probably unrealistic to assume 0 error/variance. Should I revert?

pllim commented:
https://github.com/spacetelescope/RegressionTests/actions/runs/14406105552/attempts/1#summary-40403084134

SOSS failures still there but differences seem small enough? The masterbg one will need some expert eyes.

Thanks!

melanieclarke reviewed (commented):
I still don't see why we would be seeing differences in the SOSS products - maybe it's a precision thing with some of the pre-computed squares?  Can you try to isolate which change that might be coming from?  I can look into expected values for the masterbg product.

melanieclarke commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,25 @@ def apply_flat_field(science, flat, inverse=False):
-        science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
-
         # Set the output ERR to be the combined input ERR plus flatfield ERR, summed in quadrature
         science.err = np.sqrt(science.err**2 + science.var_flat)
```
Here's the ticket to follow up on the error propagation for guider data:
https://jira.stsci.edu/browse/JP-3984

pllim commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -265,13 +264,28 @@ def test_nirspec_fs_flat(srctype):
+    assert_allclose(rt_data.slits[0].var_rnoise[nn], orig_data.slits[0].var_rnoise[nn])
+    # varflat and err not same as original input due to how they are calculated.
+    rt_varflat = (rt_data.slits[0].data[nn] * flat_val / flat_err) ** 2  # orig_input has no var_flat
+    rt_err = np.sqrt(rt_data.slits[0].var_poisson[nn] + rt_data.slits[0].var_rnoise[nn] + rt_varflat)
```
OK thanks. Good catch! I set it to None on inverse step. However in the test, it does not come back as None, but all zeroes. Does that make sense to you?

melanieclarke replied:
Yes, if you check for the data, the datamodel will automatically fill it in with the default value (0.0) and shape.  It's okay if var_flat is present and all zero after the inverse, rather than None.

melanieclarke commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -265,13 +264,27 @@ def test_nirspec_fs_flat(srctype):
+    assert_allclose(rt_data.slits[0].var_poisson[nn], orig_data.slits[0].var_poisson[nn])
+    assert_allclose(rt_data.slits[0].var_rnoise[nn], orig_data.slits[0].var_rnoise[nn])
+    assert_allclose(rt_data.slits[0].var_flat, 0)  # orig_input has no var_flat
+    # err not same as original input due to how it is calculated.
```
I think the err should now be the same - can you check?

pllim replied:
I did. It was not. If this must roundtrip back to input ERR, then I need to ponder the forward transform math a bit more.

melanieclarke replied:
Oh, it might just be because of how we set up the synthetic data - if the err array on input is not the sum in quadrature for the var components, then it won't match up when inversed.

pllim replied:
Ah yes, that makes sense. It is the same now after I fix the input ERR math. Thanks!

pllim commented:
I am pondering the NIRISS SOSS regtest errors. It makes no sense on why variance math change would affect EXTRACT1D FLUX col. I am guessing it comes from `apply_flat_field` function (rather than the other `nirspec` ones that I touched) but as suspected, changing the squared variable from `x * x` back to `x ** 2` does not fix the failure.

melanieclarke commented:
> I am pondering the NIRISS SOSS regtest errors. It makes no sense on why variance math change would affect EXTRACT1D FLUX col. I am guessing it comes from `apply_flat_field` function (rather than the other `nirspec` ones that I touched) but as suspected, changing the squared variable from `x * x` back to `x ** 2` does not fix the failure.

Can you check the photom change?  NIRISS SOSS calibrates flux after extraction, unlike the other instruments/modes.

pllim commented:
Reverting all the `x * x` to `x ** 2` in photom didn't make any difference either. I think I need to start a new branch and reapply the changes bit by bit and see, because it is passing on main. The tests are very slow, so it might be a while before I can report back. Thank you for your patience.

melanieclarke commented:
> The tests are very slow, so it might be a while before I can report back. Thank you for your patience.

No rush!  Thanks for your diligence.

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,28 @@ def apply_flat_field(science, flat, inverse=False):
+    flat_data_squared = flat_data * flat_data
     if not inverse:
         science.data /= flat_data
+        # NOTE: Re-arranging var_flat math will cause some NIRISS SOSS
+        # regression tests to fail with floating point differences.
+        science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
```
I think I found the culprit for NIRISS failures. 😭 

RT running at https://github.com/spacetelescope/RegressionTests/actions/runs/14477166563

pllim commented:
RT only has 1 failure now: `jwst.regtest.test_nirspec_masterbackground::test_nirspec_mos_mbkg[masterbg2d]`

melanieclarke commented:
> RT only has 1 failure now: `jwst.regtest.test_nirspec_masterbackground::test_nirspec_mos_mbkg[masterbg2d]`

Great, that's the only one expected.  I'll take a look at it.

melanieclarke commented on jwst/flatfield/flat_field.py:
```diff
@@ -205,23 +206,28 @@ def apply_flat_field(science, flat, inverse=False):
+        science.var_flat = science.data ** 2 / flat_data_squared * flat_err ** 2
     else:
         science.data *= flat_data
+        science.var_flat = None  # Does not exist before flatfield step
```
Instead of None, here and elsewhere:
```suggestion
        science.var_flat = np.zeros_like(science.data)  # Does not exist before flatfield step
```

melanieclarke commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -265,13 +264,25 @@ def test_nirspec_fs_flat(srctype):
+    assert_allclose(rt_data.slits[0].data[nn], orig_data.slits[0].data[nn])
+    assert_allclose(rt_data.slits[0].var_poisson[nn], orig_data.slits[0].var_poisson[nn])
+    assert_allclose(rt_data.slits[0].var_rnoise[nn], orig_data.slits[0].var_rnoise[nn])
+    assert_allclose(rt_data.slits[0].var_flat, 0)  # orig_input has no var_flat
```
```suggestion
    assert_allclose(rt_data.slits[0].var_flat[nn], 0)  # orig_input has no var_flat
```

melanieclarke reviewed (commented):
Okay, I looked at the masterbg2d output and I think it looks right.  The S/N with the updated error is comparable to what it should be, looking at the extracted spectra it's derived from.   The error and variance arrays also no longer show the influence of the var_flat array, and err is the sum in quadrature of the read noise and poisson variance. The old version had much higher error and variance values, making the S/N way off for the master bg product.

One thing: setting the var_flat to None makes it now write the extension out as a zero-size array, which behaves badly with some FITS readers.  I'm sorry to make you change this again, but I think it would be clearer/easier on users to write out zeros for the var_flat values instead of an empty array.  This will need one more change to the unit test as well, because if the var_flat extension matches the data size, it will update NaN values to match.

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -419,18 +430,21 @@ def nirspec_fs_msa(output_model, f_flat_model, s_flat_model, d_flat_model, dispa
-        slit.err = np.sqrt(slit.var_poisson + slit.var_rnoise + slit.var_flat)
+            slit.var_poisson *= flat_data_squared
+            slit.var_rnoise *= flat_data_squared
+            slit.var_flat = None  # Does not exist before flatfield step
```
```suggestion
            slit.var_flat = np.zeros_like(slit.data)  # Does not exist before flatfield step
```

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -498,21 +513,26 @@ def nirspec_brightobj(output_model, f_flat_model, s_flat_model, d_flat_model, di
         output_model.data *= interpolated_flat.data
+        output_model.var_poisson *= flat_data_squared
+        output_model.var_rnoise *= flat_data_squared
+        output_model.var_flat = None  # Does not exist before flatfield step
```
```suggestion
        output_model.var_flat = np.zeros_like(output_model.data)  # Does not exist before flatfield step
```

pllim commented on jwst/flatfield/flat_field.py:
```diff
@@ -567,21 +588,26 @@ def nirspec_ifu(output_model, f_flat_model, s_flat_model, d_flat_model, dispaxis
             output_model.data *= flat
+            output_model.var_poisson *= flat_data_squared
+            output_model.var_rnoise *= flat_data_squared
+            output_model.var_flat = None  # Does not exist before flatfield step
```
```suggestion
            output_model.var_flat = np.zeros_like(output_model.data)  # Does not exist before flatfield step
```

pllim reviewed (commented):
This is just so I can commit all the suggestions at once and keep Melanie as co-author via the GH UI.

pllim commented:
I was wondering about the FITSDiff complaining about VAR_FLAT dimensions. Thanks for the suggestions! I think I have addressed your comments and also rerunning RT again at https://github.com/spacetelescope/RegressionTests/actions/runs/14479420182
